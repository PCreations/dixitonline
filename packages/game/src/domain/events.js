export const newGameCreatedEvent = ({ gameId }) => ({
  type: '[lobby] - a new game has been created',
  payload: {
    gameId,
  },
});

export const playerJoinedGame = ({ gameId, playerId }) => ({
  type: '[lobby] - a new player has joined a game',
  payload: {
    gameId,
    playerId,
  },
});

export const newGameStartedEvent = ({ gameId, playerIds }) => ({
  type: '[lobby] - a new game has started',
  payload: {
    gameId,
    playerIds,
  },
});
