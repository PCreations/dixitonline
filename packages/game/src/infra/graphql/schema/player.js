import { objectType } from 'nexus';

export const Player = objectType({
  name: 'GamePlayer',
  definition(t) {
    t.id('id');
    t.string('name');
    t.int('score', {
      resolve(player) {
        return player.score ?? 0;
      },
    });
  },
});
