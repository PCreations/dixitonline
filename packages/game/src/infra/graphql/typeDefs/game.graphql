type Game {
  id: ID!
  host: GamePlayer!
  players: [GamePlayer!]!
}
