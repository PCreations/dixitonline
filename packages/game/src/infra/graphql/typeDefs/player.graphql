type GamePlayer {
  id: ID!
  name: String!
}
