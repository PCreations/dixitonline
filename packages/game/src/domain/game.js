/* eslint-disable max-classes-per-file */
import { shuffle as shuffleWithSeed } from 'shuffle-seed';
import {
  newGameCreatedEvent,
  playerJoinedGame,
  newGameStartedEvent,
  handsCompletedEvent,
  gameEndedEvent,
} from './events';
import { getEndingConditionStrategy } from './end-condition-strategies';
import { equals as playerEquals } from './player';
import { makeGameResult as baseMakeGameResult, makeErrorResult } from './game-result';

const defaultShuffle = toShuffle => shuffleWithSeed(toShuffle, 'default-seed');

export const DEFAULT_END_CONDITION = {
  isGameEnded: false,
};

export const CARDS_NOT_DEALT_YET = null;

export const GameStatus = {
  WAITING_FOR_PLAYERS: 'WAITING_FOR_PLAYERS',
  STARTED: 'STARTED',
  ENDED: 'ENDED',
};

export const GameError = {
  GAME_ALREADY_JOINED: 'GAME_ALREADY_JOINED',
  GAME_ALREADY_STARTED: 'GAME_ALREADY_STARTED',
  MAXIMUM_NUMBER_OF_PLAYERS_REACHED: 'MAXIMUM_NUMBER_OF_PLAYERS_REACHED',
  ONLY_HOST_CAN_START_GAME: 'ONLY_HOST_CAN_START_GAME',
  NOT_ENOUGH_PLAYERS: 'NOT_ENOUGH_PLAYERS',
  X_TIMES_STORYTELLER_CANT_BE_LESS_THAN_ONE: 'X_TIMES_STORYTELLER_CANT_BE_LESS_THAN_ONE',
  SCORE_LIMIT_CANT_BE_LESS_THAN_ONE: 'SCORE_LIMIT_CANT_BE_LESS_THAN_ONE',
};

export const MAXIMUM_NUMBER_OF_PLAYERS = 6;
export const MINIMUM_NUMBER_OF_PLAYERS = 3;
export const NUMBER_OF_CARDS_IN_A_DECK = 84;
export const NUMBER_OF_CARDS_BY_HAND = 6;

export const makeNullCards = () => ({
  isNullCards: true,
  length: 0,
});

export const makeGame = ({
  id,
  host,
  cards = makeNullCards(),
  drawPile = [],
  score = {},
  players = [],
  status = GameStatus.WAITING_FOR_PLAYERS,
  currentTurn = { id: null, storytellerId: null, number: 0 },
  endCondition = DEFAULT_END_CONDITION,
} = {}) => {
  if (!id) throw new Error('Game must contain an id');
  if (!host) throw new Error('Game must have an host');
  if (
    typeof endCondition.xTimesStorytellerLimit === 'undefined' &&
    typeof endCondition.scoreLimit === 'undefined' &&
    typeof endCondition.isGameEnded === 'undefined'
  )
    throw new Error(`Invalid end condition, received "${endCondition}"`);

  console.log(`creating a game with ${cards.length} cards`);

  return Object.freeze({
    id,
    host,
    players,
    cards,
    drawPile,
    endCondition,
    score,
    status,
    currentTurn,
  });
};

const haveCardsNotBeenDealtYet = game => game.cards.isNullCards;

const applyEndingConditionStrategy = ({ value: game, events }) => {
  if (haveCardsNotBeenDealtYet(game)) return { value: game, events };
  const endingCondition = getEndingConditionStrategy(game);
  const { isGameEnded } = endingCondition(game);
  if (isGameEnded)
    return {
      value: makeGame({ ...game, status: GameStatus.ENDED }),
      events: events.concat(gameEndedEvent({ gameId: game.id })),
    };
  return { value: game, events };
};

const makeGameResult = (game, events = []) => applyEndingConditionStrategy(baseMakeGameResult(game, events));

export const getNumberOfCardsByHand = game =>
  getAllPlayers(game).length === 3 ? NUMBER_OF_CARDS_BY_HAND + 1 : NUMBER_OF_CARDS_BY_HAND;

export const getAllPlayers = game => [game.host, ...game.players];

export const getEndCondition = game => {
  const allPlayersLength = getAllPlayers(game).length;
  if (typeof game.endCondition.xTimesStorytellerLimit !== 'undefined') {
    return {
      remainingTurns: allPlayersLength * game.endCondition.xTimesStorytellerLimit - game.currentTurn.number,
    };
  }
  if (typeof game.endCondition.scoreLimit !== 'undefined') {
    return {
      scoreLimit: game.endCondition.scoreLimit,
    };
  }
  return {
    remainingTurns: Math.floor(game.cards.length / allPlayersLength),
  };
};

const isGameFull = game => getAllPlayers(game).length === MAXIMUM_NUMBER_OF_PLAYERS;

export const createGame = ({ gameId, host, endCondition }) => {
  if (endCondition?.xTimesStorytellerLimit !== undefined && endCondition.xTimesStorytellerLimit < 1) {
    return makeErrorResult(GameError.X_TIMES_STORYTELLER_CANT_BE_LESS_THAN_ONE);
  }
  if (endCondition?.scoreLimit !== undefined && endCondition.scoreLimit < 1) {
    return makeErrorResult(GameError.SCORE_LIMIT_CANT_BE_LESS_THAN_ONE);
  }
  return makeGameResult(makeGame({ id: gameId, host, endCondition }), [newGameCreatedEvent({ gameId })]);
};

export const joinPlayer = (game, player) => {
  if (getAllPlayers(game).some(playerEquals.bind(null, player))) return makeErrorResult(GameError.GAME_ALREADY_JOINED);
  if (isGameFull(game)) return makeErrorResult(GameError.MAXIMUM_NUMBER_OF_PLAYERS_REACHED);
  return makeGameResult(makeGame({ ...game, players: game.players.concat(player) }), [
    playerJoinedGame({ gameId: game.id, playerId: player.id }),
  ]);
};

export const startGame = (game, player) => {
  if (game.status === GameStatus.STARTED) {
    return makeErrorResult(GameError.GAME_ALREADY_STARTED);
  }
  if (getAllPlayers(game).length < MINIMUM_NUMBER_OF_PLAYERS) {
    return makeErrorResult(GameError.NOT_ENOUGH_PLAYERS);
  }
  if (playerEquals(game.host, player)) {
    const isDefaultEndCondition =
      game.endCondition?.xTimesStorytellerLimit === undefined && game.endCondition?.scoreLimit === undefined;
    return makeGameResult(
      makeGame({
        ...game,
        status: GameStatus.STARTED,
      }),
      [
        newGameStartedEvent({
          gameId: game.id,
          playerIds: getAllPlayers(game).map(({ id }) => id),
          useAllDeck: !isDefaultEndCondition,
        }),
      ]
    );
  }
  return makeErrorResult(GameError.ONLY_HOST_CAN_START_GAME);
};

export const updateDeck = (game, { discardedCards, shuffle = defaultShuffle } = {}) => {
  if (game.cards.length < getAllPlayers(game).length && game.endCondition !== DEFAULT_END_CONDITION) {
    const drawPile = shuffle([...game.drawPile, ...discardedCards, ...game.cards]);
    return makeGameResult(
      makeGame({
        ...game,
        cards: drawPile,
        drawPile: [],
      })
    );
  }
  return makeGameResult(
    makeGame({
      ...game,
      drawPile: game.drawPile.concat(discardedCards),
    })
  );
};

export const completeHands = (game, { cards, actualHandsByPlayerId, previousTurnId }) => {
  const allPlayers = getAllPlayers(game);
  const numberOfCardsByHand = getNumberOfCardsByHand(game);
  const actualCards = cards ?? game.cards;
  if (!actualHandsByPlayerId) {
    const handsByPlayerId = allPlayers.reduce(
      (hands, player, playerIndex) => ({
        ...hands,
        [player.id]: actualCards.slice(
          playerIndex * numberOfCardsByHand,
          playerIndex * numberOfCardsByHand + numberOfCardsByHand
        ),
      }),
      {}
    );
    const remainingCards = actualCards.slice(allPlayers.length * numberOfCardsByHand);
    return makeGameResult(
      makeGame({
        ...game,
        cards: remainingCards,
      }),
      [handsCompletedEvent({ gameId: game.id, handsByPlayerId })]
    );
  }
  const handsByPlayerId = Object.entries(actualHandsByPlayerId).reduce(
    (completedHands, [playerId, hand], playerIndex) => {
      const numberOfCardsToCompleteHand = numberOfCardsByHand - hand.length;
      return {
        ...completedHands,
        [playerId]: hand.concat(actualCards.slice(playerIndex, playerIndex + numberOfCardsToCompleteHand)),
      };
    },
    {}
  );
  return makeGameResult(
    makeGame({
      ...game,
      cards: actualCards.slice(allPlayers.length),
    }),
    [handsCompletedEvent({ gameId: game.id, handsByPlayerId, previousTurnId })]
  );
};

export const updateScore = (game, turnScore) => {
  const newScore = Object.entries(turnScore).reduce(
    (scores, [playerId, playerTurnScore]) => ({
      ...scores,
      [playerId]: playerTurnScore + (game.score[playerId] || 0),
    }),
    {}
  );
  console.log('updated score', newScore);
  return makeGameResult(
    makeGame({
      ...game,
      score: newScore,
    })
  );
};

export const setCurrentTurn = (game, currentTurn) =>
  makeGameResult(
    makeGame({
      ...game,
      currentTurn: {
        ...currentTurn,
        number: game.currentTurn.number + 1,
      },
    })
  );
