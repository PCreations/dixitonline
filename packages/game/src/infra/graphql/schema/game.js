import { objectType, enumType, unionType } from 'nexus';
import { Player } from './player';
import { getAllPlayers, getEndCondition, GameStatus as DomainGameStatus } from '../../../domain/game';

export const GameStatus = enumType({
  name: 'GameStatus',
  members: DomainGameStatus,
});

export const RemainingTurnsEndCondition = objectType({
  name: 'GameRemainingTurnsEndCondition',
  definition(t) {
    t.int('remainingTurns');
  },
});

export const ScoreLimitEndCondition = objectType({
  name: 'GameScoreLimitEndCondition',
  definition(t) {
    t.int('scoreLimit');
  },
});

export const GameEndCondition = unionType({
  name: 'GameEndCondition',
  definition(t) {
    t.members(RemainingTurnsEndCondition, ScoreLimitEndCondition);
    t.resolveType(obj =>
      typeof obj.remainingTurns === 'undefined' ? 'GameScoreLimitEndCondition' : 'GameRemainingTurnsEndCondition'
    );
  },
});

export const Game = objectType({
  name: 'Game',
  definition(t) {
    t.id('id');
    t.field('host', { type: Player });
    t.field('status', { type: GameStatus });
    t.id('currentTurnId', {
      nullable: true,
      resolve({ currentTurn }) {
        return currentTurn.id;
      },
    });
    t.int('remainingTurns', {
      deprecation: 'use endCondition instead',
    });
    t.field('endCondition', {
      type: GameEndCondition,
      resolve(game) {
        console.log(getEndCondition(game));
        return getEndCondition(game);
      },
    });
    t.list.field('players', {
      type: Player,
      resolve(game) {
        const { score } = game;
        const players = getAllPlayers(game).map(player => {
          return {
            ...player,
            score: score[player.id],
          };
        });
        return players;
      },
    });
  },
});
