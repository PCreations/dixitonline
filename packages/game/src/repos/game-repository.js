import shortid from 'shortid';
import { makeGame, GameStatus } from '../domain/game';

export class GameNotFoundError extends Error {
  constructor(gameId) {
    super();
    this.message = `Game ${gameId} not found`;
  }
}

export const makeGameRepository = ({ uuid = shortid, firestore = makeNullFirestore() } = {}) => {
  const lobbyGames = firestore.collection('lobby-games');
  return {
    getNextGameId() {
      return uuid();
    },
    saveGame(game) {
      return lobbyGames.doc(game.id).set(game);
    },
    async getGameById(id) {
      const doc = await lobbyGames.doc(id).get();
      if (!doc.exists) {
        throw new GameNotFoundError(id);
      }
      return makeGame(doc.data());
    },
    getAllGames() {
      return lobbyGames
        .where('status', '==', GameStatus.WAITING_FOR_PLAYERS)
        .get()
        .then(snapshot => {
          const games = [];
          snapshot.forEach(doc => games.push(makeGame(doc.data())));
          return games;
        });
    },
    deleteGameById(gameId) {
      return lobbyGames.doc(gameId).delete();
    },
  };
};

export const makeNullGameRepository = ({ nextGameId, gamesData }) =>
  makeGameRepository({ uuid: nullUuid(nextGameId), firestore: makeNullFirestore(gamesData) });

const nullUuid = expectedUuid => () => expectedUuid;

const makeNullFirestore = (gamesInitialData = {}) => {
  const gamesData = { ...gamesInitialData };
  return {
    collection() {
      return {
        where(field, _, value) {
          return {
            async get() {
              const docs = Object.values(gamesData)
                .filter(Boolean)
                .filter(game => game[field] === value)
                .map(game => ({
                  data() {
                    return game;
                  },
                }));
              return {
                forEach: docs.forEach.bind(docs),
              };
            },
          };
        },
        doc(gameId) {
          return {
            get() {
              return {
                data() {
                  return gamesData[gameId];
                },
                exists: typeof gamesData[gameId] !== 'undefined',
              };
            },
            async set(game) {
              gamesData[gameId] = game;
            },
            delete() {
              gamesData[gameId] = undefined;
            },
          };
        },
      };
    },
  };
};
