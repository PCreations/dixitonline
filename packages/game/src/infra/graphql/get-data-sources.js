import { makeGameRepository } from '../../repos/game-repository';

export const makeGetDataSources = ({ gameRepository = makeGameRepository() } = {}) => () => ({
  gameRepository,
});
