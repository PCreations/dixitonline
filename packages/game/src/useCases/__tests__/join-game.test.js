import gql from 'graphql-tag';
import { createTestClient } from 'apollo-server-testing';
import { makeNullGameRepository } from '../../repos/game-repository';
import { makeTestServer } from '../../__tests__/test-server';
import { buildgameRepositoryInitialGames } from '../../__tests__/dataBuilders/game-repository-initial-games';
import { buildTestGame } from '../../__tests__/dataBuilders/game';
import { buildTestPlayer } from '../../__tests__/dataBuilders/player';
import { makeGetDataSources } from '../../infra/graphql/get-data-sources';
import { playerJoinedGame } from '../../domain/events';

const getJoinGameTestServer = ({
  currentUserId = 'p2',
  currentUserUsername = 'player2',
  maxOutPlayers = false,
} = {}) => {
  const host = buildTestPlayer()
    .withId('p1')
    .withName('player1')
    .build();
  let gameToBeBuilt = buildTestGame()
    .withId('g1')
    .withHost(host);
  if (maxOutPlayers) {
    gameToBeBuilt = gameToBeBuilt.asFullGame();
  }
  const game = gameToBeBuilt.build();
  const initialGames = buildgameRepositoryInitialGames()
    .withGames([game])
    .build();
  const gameRepository = makeNullGameRepository({
    gamesData: initialGames,
  });
  const dispatchDomainEvents = jest.fn();
  const server = makeTestServer({
    getDataSources: makeGetDataSources({
      gameRepository,
    }),
    dispatchDomainEvents,
    currentUserId,
    currentUserUsername,
  });

  return {
    game,
    dispatchDomainEvents,
    gameRepository,
    server,
  };
};

describe('join game', () => {
  test('a player can join a game', async () => {
    // arrange
    const { game, dispatchDomainEvents, gameRepository, server } = getJoinGameTestServer();
    const expectedUpdatedGame = buildTestGame(game)
      .withPlayers([
        buildTestPlayer()
          .withId('p2')
          .withName('player2')
          .build(),
      ])
      .build();
    const GAME_JOIN_GAME = gql`
      mutation GameJoinGame($joinGameInput: GameJoinGameInput!) {
        gameJoinGame(joinGameInput: $joinGameInput) {
          ... on GameJoinGameResultSuccess {
            game {
              id
              host {
                id
                name
              }
              players {
                id
                name
              }
            }
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: GAME_JOIN_GAME,
      variables: { joinGameInput: { gameId: 'g1' } },
      operationName: 'GameJoinGame',
    });
    const updatedGame = await gameRepository.getGameById('g1');

    // assert
    expect(response).toMatchSnapshot();
    // TODO : change this when https://github.com/facebook/jest/pull/9575 is released
    expect({ ...updatedGame }).toEqual({ ...expectedUpdatedGame });
    expect(dispatchDomainEvents).toHaveBeenCalledWith([playerJoinedGame({ gameId: 'g1', playerId: 'p2' })]);
  });
  test("a player can't join a game she has already joined", async () => {
    // arrange
    const { game, dispatchDomainEvents, gameRepository, server } = getJoinGameTestServer({
      currentUserUsername: 'player1',
      currentUserId: 'p1',
    });
    const GAME_JOIN_GAME = gql`
      mutation GameCreateGame($joinGameInput: GameJoinGameInput!) {
        gameJoinGame(joinGameInput: $joinGameInput) {
          ... on GameJoinGameResultError {
            type
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: GAME_JOIN_GAME,
      variables: {
        joinGameInput: { gameId: 'g1' },
      },
      operationName: 'GameCreateGame',
    });
    const updatedGame = await gameRepository.getGameById('g1');

    // assert
    expect(response).toMatchSnapshot();
    // TODO : change this when https://github.com/facebook/jest/pull/9575 is released
    expect({ ...updatedGame }).toEqual({
      ...game,
    });
    expect(dispatchDomainEvents).not.toHaveBeenCalled();
  });
  test("a player can't join a game with already the maximum number of player", async () => {
    // arrange
    const { game, dispatchDomainEvents, gameRepository, server } = getJoinGameTestServer({ maxOutPlayers: true });
    const GAME_JOIN_GAME = gql`
      mutation GameCreateGame($joinGameInput: GameJoinGameInput!) {
        gameJoinGame(joinGameInput: $joinGameInput) {
          ... on GameJoinGameResultError {
            type
          }
        }
      }
    `;

    // act
    const { mutate } = createTestClient(server);
    const response = await mutate({
      mutation: GAME_JOIN_GAME,
      variables: {
        joinGameInput: { gameId: 'g1' },
      },
      operationName: 'GameCreateGame',
    });
    const updatedGame = await gameRepository.getGameById('g1');

    // assert
    expect(response).toMatchSnapshot();
    // TODO : change this when https://github.com/facebook/jest/pull/9575 is released
    expect({ ...updatedGame }).toEqual({
      ...game,
    });
    expect(dispatchDomainEvents).not.toHaveBeenCalled();
  });
  test.todo("a player can't join an already started game");
});
