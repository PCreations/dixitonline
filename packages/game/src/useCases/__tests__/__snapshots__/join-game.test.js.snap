// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`join game a player can join a game 1`] = `
Object {
  "data": Object {
    "gameJoinGame": Object {
      "game": Object {
        "host": Object {
          "id": "p1",
          "name": "player1",
        },
        "id": "g1",
        "players": Array [
          Object {
            "id": "p1",
            "name": "player1",
          },
          Object {
            "id": "p2",
            "name": "player2",
          },
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`join game a player can't join a game she has already joined 1`] = `
Object {
  "data": Object {
    "gameJoinGame": Object {
      "type": "GAME_ALREADY_JOINED",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`join game a player can't join a game with already the maximum number of player 1`] = `
Object {
  "data": Object {
    "gameJoinGame": Object {
      "type": "MAXIMUM_NUMBER_OF_PLAYERS_REACHED",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
